### ISSUE TITLE
Chrome displays 24 instead of 00

### ISSUE THREAD

**Issue by kwtucker** [moment/luxon#726]
`DateTime.fromFormat("2020-06-12 00:00", "yyyy-MM-dd T").toFormat('T')`
**Result**: 24:00
**Expected**: 00:00

**Comment by GillesDebunne** [comment]
I'm unable to reproduce this problem. On which browser(s) or node version(s) does it happen?

**Comment by thomas-darling** [comment]
We encountered this too - it's because the locale data for the Intl API in the browser is wrong for some locales.
And it's worse than it appears - the time `00:15` is formatted as `24:15` too, which is just stupid.

I don't remember all the OS, browsers and locale combinations for which this occurs, but I just tested it again, and it definitely still occurs in Chrome 84 on macOS Catalina for the `en-US` locale. Here's a repro using the native API:

```ts
const formatOptions = { hour: 'numeric', minute: 'numeric', hour12: false };
const valueToFormat = Date.parse('2020-01-01T00:15:00');

const dateTimeFormat = new Intl.DateTimeFormat('en-US', formatOptions);
console.log(dateTimeFormat.format(valueToFormat));

// Outputs "24:15"
```

This really should be fixed in the browsers, but in the absence of that, it would be great if Luxon could include a fix.

**Comment by GillesDebunne** [comment]
**TL;DR;**

Chrome v80 introduced support for the `hourCycle` option in `Intl.DateTimeFormat`, which specifies if midnight should be displayed as `00` or `24`. However, locales using a _12 hour_ format by default (such as `en-US`) now display midnight as `24` when asked to use _24 hour_ instead. This is a bug in the ECMA specifications, which should be fixed in ECMA2021.

**Details**

The root cause of this issue comes from [this change in Chromium](https://chromium.googlesource.com/v8/v8/+/07437481c7fb47d7e84e4ce04008339d817e08ce). It introduces support for the `hourCycle` option in `Intl.DateTimeFormat`.

The four supported values are _"h11", "h12", "h23", or "h24"_ ([MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/DateTimeFormat/DateTimeFormat)) and the difference between `h23` and `h24` is precisely if midnight should be displayed as `00` (_h23_) or `24` (_h24_) ([details](https://stackoverflow.com/a/55823036)).

However, Luxon uses the `hour12: false` option instead, which _overrides the hourCycle option in case both are present_ ([MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/DateTimeFormat/DateTimeFormat)). In that case, deciding which of `h23` or `h24` to use in defined in this [ECMA DateTime specification](https://ecma-international.org/ecma-402/#sec-initializedatetimeformat) at §36, and depends on the _locale's default format_. For the _"en-US"_ locale, the default 12 hour AM/PM format (`h12`) results in `h24` when requesting a 24 hour display.

Note that **no region defines `h24` has an allowed format** and this clearly is a bug in the ECMA specifications, which should be fixed by [this PR](https://github.com/tc39/ecma402/pull/436) in the 2021 version. A Chrome [bug report](https://support.google.com/chrome/thread/29828561) also tracks this issue.

**Workaround**

A workaround could be to use `hourCycle: 'h23'` instead of `hour12: false` in Luxon. However, since this option might not be supported, we should first run a fake format to detect if it is, and change all the formats accordingly?

**Comment by kwtucker** [comment]
Another workaround is if you change the format from "yyyy-MM-dd  T" to "yyyy-MM-dd  HH:mm".

`DateTime.fromFormat("2020-06-12 00:00", "yyyy-MM-dd HH:mm").toFormat('HH:mm')`
Result: 00:00

**Comment by balagge** [comment]
Strangely, I get `"0:00"` (not `"00:00"`)

Ubuntu 14
Node v12.18.3
luxon@1.25.0

**Comment by joshland96** [comment]
I'm seeing this as well on my local machine.

```js
console.log(dateTime.toString())
console.log(dateTime.toLocaleString(DateTime.TIME_24_SIMPLE))

// Output:
// 2021-01-05T00:00:00.000Z
// 24:00
```

MacOS: 11.2.3
Node: 12.19.11
Luxon: 1.26.0

**Comment by austinbutler** [comment]
From what I can tell there's no path forward:

- The [Chromium bug](https://bugs.chromium.org/p/chromium/issues/detail?id=1045791) has been closed as "WontFix"
- The [ECMA PR](https://github.com/tc39/ecma402/pull/436) mentioned above was closed without anything being done/merged
- There's only an [open discussion](https://github.com/tc39/ecma402/issues/402) in the ECMA repo from what I can tell, no concrete plans

@GillesDebunne you mentioned the bug had been fixed on ECMA's side, do you have a reference for that by chance?

Are there any plans to work around this on Luxon's end?

**Comment by icambron** [comment]
We had a few instances of this issue, and I guess I loost track of this one.

@austinbutler This is fixed in Luxon 2.x, which uses `hourCycle: h23` instead of `hourCycle: false` in the 24-hour presets. This does mean that older browsers lacking `hourCycle` support won't get 24-hour times from those presets but at this point it's the better tradeoff, because all the supported Node versions and all the major browsers support it, and of course they were getting terrible behavior from these presets.

**Comment by austinbutler** [comment]
@icambron maybe I'm just hitting another bug or simply doing it wrong then. I'm still able to see this issue with Luxon 2.1.1 and Chrome 95.0.4638.69: https://codesandbox.io/s/elated-bohr-sq57i

**Comment by austinbutler** [comment]
Actually I think I'm hitting some type bug. This seems to do what I want, but it won't compile without the `@ts-ignore`: https://codesandbox.io/s/sharp-violet-8kn3m

**Comment by icambron** [comment]
@austinbutler Typescript typing aside, your code seems to work fine? I clicked on the link and fiddled with the dates and it all looked good to me

**Comment by austinbutler** [comment]
@icambron https://codesandbox.io/s/elated-bohr-sq57i gives me `September 15, 2021, 24:30 PDT` while https://codesandbox.io/s/sharp-violet-8kn3m gives me `September 15, 2021, 00:30 PDT`.

**Comment by icambron** [comment]
@austinbutler Yes, I meant your second link. Luxon can't change the behavior of `hour12`; that's defined by the browser and is subject to the spec bug discussed above. And it takes precedence in the Intl call over `hourCycle`. You should always use _only_ `hourCycle`. That's the fix Luxon made in the 24-hour presets. For example, `TIME_24_SIMPLE` is defined as:

```js
{hour: 'numeric', minute: 'numeric', hourCycle: 'h23'}
```

If Luxon had a DATETIME_FULL_24, it would be defined as `{...DATETIME_FULL, hourCycle: "h23"}`

**Comment by austinbutler** [comment]
Thanks @icambron, the typing issue lead me in the wrong direction, but I think we're on the same page now.

**Comment by Abhinavmod** [comment]
@austinbutler @icambron  When I try to run this :   const testData3 = LuxonDateTime.fromSeconds(1642832200,{zone: "America/Chicago"}); 

I get an offset of 1080  this happens only for  12 AM  How can fix this?
